### PR TITLE
ClickHouse M1 Docker

### PR DESCRIPTION
**Pull request by mariusandra**
## Ready for review

Skip the noise and scroll to the final comment: https://github.com/PostHog/posthog/pull/5215#issuecomment-917242765

## Changes

After adapting and migrating dockerfiles for `clickhouse` and `kafka`, I can now get this far on my M1 Air:

- `docker-compose -f ee/docker-compose.ch.arm64.yml up clickhouse kafka zookeeper`
- `python manage.py migrate_clickhouse`

Screenshot 2021-07-19 at 23 39 37

There's something wrong and not all tables were created, even though the migration script said it was successful. I'm not sure why that is... and will check later.

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Comment by mariusandra** [comment]
So the `events` table is not created. I believe I'm running across [this fix](https://github.com/ClickHouse/ClickHouse/pull/26256) in a very recent version of Clickhouse, as this is what I get when I run the `CREATE TABLE events` SQL manually:

> Code: 59. DB::Exception: There was an error on [localhost:9000]: Code: 59. DB::Exception: Invalid sampling column type in storage parameters: UUID. Must be one unsigned integer type. (ILLEGAL_TYPE_OF_COLUMN_FOR_FILTER) (version 21.9.1.7509 (official build)). (ILLEGAL_TYPE_OF_COLUMN_FOR_FILTER) (version 21.9.1.7509 (official build))

If I remove the `SAMPLE BY uuid` [line from here](https://github.com/PostHog/posthog/blob/401268bdba18663dd8100d26df584d420bbd2692/ee/clickhouse/sql/events.py#L38), the query finishes and the events table was created. 

@macobo do you have any context if this line is needed, or if we actually have a bug here with the SAMPLE line? 

Other notes for myself:
- Kafka sometimes still does a code dumps on boot, but that's rare and usually the next restart of the services solves the problem.
- I had to add `KAFKA_URL=kafka://localhost` to get django to talk to it. The setting KAFKA_HOSTS that's used by the plugin server was ignored.
- ClickHouse keeps throwing `libunwind: Unsupported .eh_frame_hdr version` into the console. Should be just [noise](https://github.com/ClickHouse/ClickHouse/issues/15638) that can be removed by disabling the query profiler.

At the end of this (and if I manually remove the `SAMPLE BY uuid` line) I have Django and the plugin server working properly with EE locally:

image

... yet the events still don't reach the `events` table. 

To be continued.

**Comment by mariusandra** [comment]
Clickhouse works on a M1 mac with these instructions, just something is up with the Kafka in between the containers. Will investigate further...

**Comment by mariusandra** [comment]
Gave it another long look, but still didn't get things running as they should. Managed to greatly simplify the setup though (no need for a custom container for Kafka), but still some issues.

1. This can only run with Clickhouse `21.9`, as that's the prebuilt aarch64 binary that yandex provides. I understand we're on 21.6 still. What's the blocker for upgrading to something greater? Arm64 support was backported to v21.7+.

2. The "migrate_clickhouse" script errors without an error when I leave the line "[SAMPLE BY uuid](https://github.com/PostHog/posthog/blob/c2bc2fecd0f094fc230d8d0d6c8f999cfe3741a1/ee/clickhouse/sql/events.py#L38)" inside the `create table events` command. All other SQL errors while migrating get thrown in the console, but not this. The `events` table will just not be created. Running the command manually gives the error: "Invalid sampling column type in storage parameters: UUID. Must be one unsigned integer type.". Is this something new from post-21.6? Do we need this line?

3. When I patch that error, no events end up in the database. Manually selecting from the kafka tables also gives nothing. The plugin server however successfully exports events to Kafka. I'm not sure where this gets broken. Tried debugging, but couldn't get anywhere. I didn't check, but perhaps this is also broken on the intel version with CH 21.9?

@macobo @EDsCODE is there anything you can suggest that could help move this forward?

**Comment by macobo** [comment]
Quick dump of answers:

1. Upgrading is a few hours of work on cloud and requires testing on charts. Not sure if any new versions provide us with any features we're missing (besides M1 support)
1. Sampling is something we do want to support eventually, not sure how easy to get rid of without re-ingesting the whole dataset (bad idea)
1. Not sure what causes it not to ingest. I'd recommend looking at `docker exec -it posthog_clickhouse_1 bash -c "cat /var/log/clickhouse-server/clickhouse-server.err.log" | less` perhaps it has something interesting?

**Comment by mariusandra** [comment]
Update:
1. When running the same container on intel (swapping out the ubuntu arch and also `aarch64` -> `amd64` in the pre built image download script), everything works (except for the sample by issue, which I commented out). The events kept coming in.
2. Checking the logs on the pod, I get something better:
```

2021.08.23 08:42:50.446095 [ 383 ] {} <Warning> StorageKafka (kafka_person_distinct_id): Can't get assignment. Will keep trying.
2021.08.23 08:42:50.628470 [ 379 ] {} <Warning> StorageKafka (kafka_session_recording_events): Can't get assignment. Will keep trying.
2021.08.23 08:42:50.801783 [ 388 ] {} <Warning> StorageKafka (kafka_person): Can't get assignment. Will keep trying.
2021.08.23 08:42:50.852789 [ 381 ] {} <Error> void DB::StorageKafka::threadFunc(size_t): Code: 73. DB::Exception: Unknown format Protobuf. (UNKNOWN_FORMAT), Stack trace (when copying this message, always include the lines below):

0. /build/build_docker/../contrib/poco/Foundation/src/Exception.cpp:28: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xdcf0c74 in /usr/bin/clickhouse
1. /build/build_docker/../src/Common/Exception.cpp:59: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x740e968 in /usr/bin/clickhouse
2. /build/build_docker/../src/Formats/FormatFactory.cpp:43: DB::FormatFactory::getCreators(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0xd4f1234 in /usr/bin/clickhouse
3. /build/build_docker/../contrib/libcxx/include/functional:2236: DB::FormatFactory::getInputFormat(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::ReadBuffer&, DB::Block const&, std::__1::shared_ptr<DB::Context const>, unsigned long, std::__1::optional<DB::FormatSettings> const&) const @ 0xd4f2438 in /usr/bin/clickhouse
4. /build/build_docker/../contrib/libcxx/include/optional:225: DB::KafkaBlockInputStream::readImpl() @ 0xd1fabd8 in /usr/bin/clickhouse
5. /build/build_docker/../src/DataStreams/IBlockInputStream.cpp:57: DB::IBlockInputStream::read() @ 0xc863c2c in /usr/bin/clickhouse
6. /build/build_docker/../contrib/libcxx/include/vector:658: DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::__1::function<void (DB::Block const&)> const&, std::__1::atomic<bool>*) @ 0xc87e870 in /usr/bin/clickhouse
7. /build/build_docker/../contrib/libcxx/include/functional:2191: DB::StorageKafka::streamToViews() @ 0xd1db8d4 in /usr/bin/clickhouse
8. /build/build_docker/../src/Storages/Kafka/StorageKafka.cpp:0: DB::StorageKafka::threadFunc(unsigned long) @ 0xd1dabec in /usr/bin/clickhouse
9. /build/build_docker/../src/Common/Stopwatch.h:12: DB::BackgroundSchedulePoolTaskInfo::execute() @ 0xca1f398 in /usr/bin/clickhouse
10. /build/build_docker/../contrib/poco/Foundation/include/Poco/AutoPtr.h:90: DB::BackgroundSchedulePool::threadFunction() @ 0xca2096c in /usr/bin/clickhouse
11. /build/build_docker/../src/Common/ThreadPool.h:183: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1&&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0xca20ef4 in /usr/bin/clickhouse
12. /build/build_docker/../contrib/libcxx/include/functional:2210: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x743fd60 in /usr/bin/clickhouse
13. /build/build_docker/../contrib/libcxx/include/memory:1655: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >(void*) @ 0x7441c38 in /usr/bin/clickhouse
14. start_thread @ 0x84fc in /usr/lib/aarch64-linux-gnu/libpthread-2.31.so
15. ? @ 0xd467c in /usr/lib/aarch64-linux-gnu/libc-2.31.so
 (version 21.9.1.7509 (official build))
```

Doesn't answer "why?", but at least this is something to work with.

**Comment by mariusandra** [comment]
I believe that for whatever reason, `Protobuf` is not included with the precompiled Aarch64 clickhouse bin. 

1. Precompiled aarch64 binary, the string `protobuf` is on 61 lines:

```
root@be947c164427:/docker-entrypoint-initdb.d# strings /usr/bin/clickhouse |grep -i Protobuf|wc
     61      69    2753
```

2. Precompiled intel binary, the string `protobuf` is on 73k lines:

```
root@a004c6d132b9:/# strings /usr/bin/clickhouse |grep -i Protobuf |wc
  73921  133401 9789832
```

This includes what clearly looks like google library code:

```
ZN2DB12_GLOBAL__N_124ProtobufSerializerNumberImE12setFunctionsEvEUlmE2_
ZN2DB12_GLOBAL__N_124ProtobufSerializerNumberImE12setFunctionsEvEUlvE5_
ZN2DB12_GLOBAL__N_124ProtobufSerializerNumberImE12setFunctionsEvEUlvE6_
...
N6google8protobuf10TextFormat6Parser10ParserImpl20ParserErrorCollectorE
N6google8protobuf10TextFormat7Printer33FastFieldValuePrinterUtf8EscapingE
N6google8protobuf10TextFormat7Printer28DebugStringFieldValuePrinterE
N6google8protobuf10TextFormat7Printer13TextGeneratorE
```

Thus the next step is manual compilation of a CH arm64 image with all the extensions that we need.

**Comment by mariusandra** [comment]
Some updates!

1. I heard from @hazzadous that he managed to get Clickhouse running locally on his M1 Mac by just compiling it from source. Probably with [these instructions](https://clickhouse.tech/docs/en/development/build-osx/). A bit complicated to set up, but it just works apparently.

2. I put in another few hours today in getting ClickHouse running inside a docker container on a M1. I got really far. The trouble is it takes ~1.5h to compile, and if there's an error at the end of the process, I need to restart the entire thing. Finally after a few false starts (wrong compiler version, etc), I got the compilation to `[7835/8120]` when I got an `LLVM ERROR: IO failure on output stream: No space left on device` error on a linker step. So I'll increase space and try again, and report back in 2h. However once this completes, I can save the built executable and distribute it as a "single binary url", which can be plugged in to any clickhouse script. I could also distribute the built container on Docker Hub.

Anyway, almost there! ClickHouse legitimately works on a M1 now, but is just slightly complicated to set up.

**Comment by mariusandra** [comment]
And it works!!!

Checkout this branch, run `docker-compose -f ee/docker-compose.ch.yml build clickhouse` and 90min later you will have this:

image

(also give docker more memory and disk space, or it'll bork)

I did have to comment out the `SAMPLE BY uuid` line in `create table events` to get PostHog to work with Clickhouse 2.9, but that's an [unrelated issue](https://github.com/PostHog/posthog/issues/5684), and something we can probably live with for now.

The generated clickhouse executable is huge though: 1.9GB. The generated docker image with all the build artefacts is even *huger*: 32.7GB

The next step is to upload the 1.9GB executable somewhere (S3?), and then use it as the `single_binary_location_url` inside non-compiled `arm64.Dockerfile`. Then we could even upload that package into docker hub for even easier usage.

Final context: Yandex does provide Aarch64 single binaries for Clickhouse, but due to cross-compilation issues, they come without support for `protobuf`, which we need to talk to Kafka.

**Comment by hazzadous** [comment]
@mariusandra yes those were the instructions

**Comment by mariusandra** [comment]
This is finally working. The Dockerfile could use cleanup and currently produces a 30GB image, but I'd hope to leave this to another PR.

The new instructions when running PostHog on an Apple Silicon machine:

1. Update your Docker Desktop conf to 8GB of RAM, plenty of CPUs and ~100GB of space.
2. Run `yarn arm64:build:clickhouse`. This builds an image `clickhouse-dev-arm64:latest` and takes ~1.5h.
3. Set Docker back to 2GB of RAM 
4. Depending on your dev flow, do any of these:
    1. run `yarn arm64:ch-dev:start` instead of `yarn ch-dev:start` 
    2. Run `yarn arm64:services:start` to start just `kafka` and `clickhouse` (and, yes, also `zookeeper`)
    3. Use the `ee/docker-compose.ch.arm64.yml` file instead of `ee/docker-compose.ch.yml`
5. Run `yarn dev:migrate:clickhouse` to create the tables

(yes, I'll update the docs if this lands)

The only bad thing is that we're running ClickHouse 21.9 for development now, but 21.6 everywhere else. There's [a know issue](https://github.com/PostHog/posthog/issues/5684) that prevents us from upgrading. Hence I had to [add the following](https://github.com/PostHog/posthog/pull/5215/files#diff-54bf83e1f7eb3c891bf07c308627f439b76bb1451dd026e40f0f61f821aded73R47) into `ee/clickhouse/sql/events.py`:

```py
sample_by_uuid="SAMPLE BY uuid" if not DEBUG else "",
```

I hope that's an acceptable workaround until the issue gets resolved properly :).

File: docker/clickhouse-builder/assets/arm64.compile.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+RUN apt-get -y install git cmake python ninja-build
+RUN git clone --branch $GIT_TAG --recursive https://github.com/ClickHouse/ClickHouse.git
+
+RUN apt-get -y install clang-12 build-essential
+ENV CC clang-12
+ENV CXX clang++-12
+
+RUN cd ClickHouse && mkdir build && cd build && cmake ..
+RUN cd ClickHouse/build && ninja -j 2
+RUN mv ClickHouse/build/programs/clickhouse* /usr/bin/
```
Comment (hazzadous) [inline docker/clickhouse-builder/assets/arm64.compile.Dockerfile:36]: Can we do a multistage build here and just copy over the binary? Might be simpler that the single binary url approach. And maybe will be simple to push up to dockerhub

File: ee/docker-compose.ch.arm64.yml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            POSTGRES_PASSWORD: posthog
+        ports:
+            - '5432:5432'
+    redis:
+        image: 'redis:alpine'
+        ports:
+            - '6379:6379'
+    clickhouse:
+        # Build with: yarn arm64:build:clickhouse
+        image: clickhouse-dev-arm64:latest
```
Comment (hazzadous) [inline ee/docker-compose.ch.arm64.yml:18]: Is there a simple way we can avoid having another docker compose file? Ie choosing an image based on arch env var, or using a multi arch image?

Comment (hazzadous) [inline ee/docker-compose.ch.arm64.yml:18]: Is there a simple way we can avoid having another docker compose file? Ie choosing an image based on arch env var, or using a multi arch image?

**Comment by mariusandra** [comment]
@hazzadous very fair questions. 

I think the actual real solution is to upgrade ClickHouse's own build platform to publish multiarch images via docker's `buildx` (emulates the build host), instead of using ClickHouses's current `llvm`-based multiarch build system, which builds an `arm64` binary on an `amd64` system, and [can't link the right protocol buffers](https://github.com/ClickHouse/ClickHouse/issues/28018#issuecomment-903638561). We could build this and contribute it upstream.

Currently ClickHouse's Dockerfiles are built by just `apt get install`ing the right prebuilt `.deb`, which comes from some other non-docker build pipeline.

Save that, we could build a pipeline like that, and publish our own multiplatform `PostHog/clickhouse` images to the Docker Hub. We'd then use these images both in the Docker compose files, and in production, for perfect version parity.

We might need to do this eventually anyway, if we want to run beta or custom builds of ClickHouse.

...

However, this PR just solves an immediate need: getting PostHog working on a M1. It's suboptimal, but works. I'd love to improve it in a separate PR. First by making a system that creates much smaller images.

File: ee/docker-compose.ch.arm64.yml (hunk 2 above)
Comment (mariusandra) [reply ee/docker-compose.ch.arm64.yml:18]: > choosing an image based on arch env
I looked, but didn't find any way to do that. I think that's intentional as well.

File: docker/clickhouse-builder/assets/arm64.compile.Dockerfile (hunk 1 above)
Comment (mariusandra) [reply docker/clickhouse-builder/assets/arm64.compile.Dockerfile:36]: We could. Two options:
1. Running `docker cp` and copying the binary out of one build container and into another production container
2. Running all the steps in this dockerfile inside one `RUN` command, which also deletes all build and installation artefacts --> the only changes in the volume will be the new build binary.

Just to be able to cache steps when building the Dockerfile (the git clone already takes 10min), I built it now as is.

Comment (hazzadous) [reply docker/clickhouse-builder/assets/arm64.compile.Dockerfile:36]: Right, yes 2 is a bit awkward with caching and all.

To clarify, I was thinking something like 

```
FROM arm64v8/ubuntu:focal AS builder
...
RUN cd ClickHouse && mkdir build && cd build && cmake ..
RUN cd ClickHouse/build && ninja -j 2

FROM arm64v8/ubuntu:focal

COPY --from=builder /.../ClickHouse/builder/programes/clickhouse* /usr/bin/

...
```

So we can keep all the caching for the build, but then have a small image to push up to dockerhub

But yeah getting this in as is would be great.

File: ee/docker-compose.ch.arm64.yml (hunk 2 above)
Comment (hazzadous) [reply ee/docker-compose.ch.arm64.yml:18]: You could do something like `image: "${CLICKHOUSE_IMAGE:-yandex/clickhouse-server:21.6.5}"` so we'd default to yandex if `CLICKHOUSE_IMAGE` isn't set. Then just set this env var in the package.json scripts for arm.

File: docker/clickhouse-builder/assets/arm64.compile.Dockerfile (hunk 1 above)
Comment (mariusandra) [reply docker/clickhouse-builder/assets/arm64.compile.Dockerfile:36]: Oh, can we do that in one dockerfile? If so, 😍

**Comment by mariusandra** [comment]
Also, feel free to directly commit some improvements into this PR :)

**Comment by hazzadous** [comment]
@mariusandra cool, yeah I'd be happy to get this in as is and then make improvements after

**Comment by hazzadous** [comment]
(read, I'm in the middle of something at the moment and don't want to hold up)

**Review by hazzadous — APPROVED** [review]

**Comment by samwinslow** [comment]
Having some trouble getting the `ee_web_1` container to work properly. I've gotten as far as being able to get some kind of HTML response from `localhost:8000`, but it doesn't seem to start Webpack correctly, as any resources on :8234 give `ERR_EMPTY_RESPONSE`.

Also tried to run `./bin/start-frontend` concurrently with the Docker containers, and although Kea Typegen and Webpack both compile their assets successfully, any requests to :8000 or :8234 would fail indefinitely.

At some point it also becomes impossible to kill or remove `ee_web_1`, and the only way I can seemingly kill it is to restart my system... any suggestions?

**Pull request merged**